From ce594ce7312eb8facffc52d20360d9cd51c6fc33 Mon Sep 17 00:00:00 2001
From: Joel Jakobsson <joel@compiler.org>
Date: Fri, 17 Feb 2023 01:26:29 +0100
Subject: [PATCH 5/6] Upgrade NumericDigit type from int16 to int32.

This is a preparatory change, to make room for larger NBASE values.
No algorithms have been changed. Types for carry and other intermediary
variables don't need to be changed, since they cannot overflow, since
the code hasn't been changed.

The jsonb_jsonpath test change is expected, since the .keyvalue() "id"
value is based on the difference between memory addresses,
and the specific test that changed contains Numeric jsonb values.
---
 src/backend/utils/adt/numeric.c       | 50 +++++++++++++++++----------
 src/test/regress/expected/numeric.out | 12 +++----
 2 files changed, 37 insertions(+), 25 deletions(-)

diff --git a/src/backend/utils/adt/numeric.c b/src/backend/utils/adt/numeric.c
index 63b4baedba..417350fb1f 100644
--- a/src/backend/utils/adt/numeric.c
+++ b/src/backend/utils/adt/numeric.c
@@ -100,7 +100,7 @@ typedef signed char NumericDigit;
 #define DIV_GUARD_DIGITS	4

 typedef uint8 NumericDigitData;
-typedef int16 NumericDigit;
+typedef int32 NumericDigit;
 #endif

 /*
@@ -492,7 +492,8 @@ static void dump_var(const char *str, NumericVar *var);
 #define NUMERIC_DIGITS(num) (NUMERIC_HEADER_IS_SHORT(num) ? \
 	(num)->choice.n_short.n_data : (num)->choice.n_long.n_data)
 #define NUMERIC_NBYTES(num) (VARSIZE(num) - NUMERIC_HEADER_SIZE(num))
-#define NUMERIC_NDIGITS(num) ((NUMERIC_NBYTES(num) + 1) / sizeof(NumericDigit))
+#define NUMERIC_NDIGITS(num) \
+	((NUMERIC_NBYTES(num) + sizeof(NumericDigit) - 1) / sizeof(NumericDigit))
 #define NUMERIC_CAN_BE_SHORT(scale,weight) \
 	((scale) <= NUMERIC_SHORT_DSCALE_MAX && \
 	(weight) <= NUMERIC_SHORT_WEIGHT_MAX && \
@@ -1066,8 +1067,9 @@ numeric_normalize(Numeric num)
 /*
  *		numeric_recv			- converts external binary format to numeric
  *
- * External format is a sequence of int16's:
- * ndigits, weight, sign, dscale, NumericDigits.
+ * External header format is a sequence of int16's:
+ * ndigits, weight, sign, dscale.
+ * followed by a sequence of bytes for NumericDigits.
  */
 Datum
 numeric_recv(PG_FUNCTION_ARGS)
@@ -1171,7 +1173,7 @@ numeric_send(PG_FUNCTION_ARGS)
 	pq_sendint16(&buf, x.sign);
 	pq_sendint16(&buf, x.dscale);
 	for (i = 0; i < x.ndigits; i++)
-		pq_sendint16(&buf, x.digits[i]);
+		pq_sendint32(&buf, x.digits[i]);

 	PG_RETURN_BYTEA_P(pq_endtypsend(&buf));
 }
@@ -2547,8 +2549,8 @@ cmp_numerics(Numeric num1, Numeric num2)
 		NumericDigit fixed_buf2[2] = {0};
 		int nbytes1 = NUMERIC_NBYTES(num1);
 		int nbytes2 = NUMERIC_NBYTES(num2);
-		int ndigits1 = (nbytes1 + 1) / sizeof(NumericDigit);
-		int ndigits2 = (nbytes2 + 1) / sizeof(NumericDigit);
+		int ndigits1 = (nbytes1 + sizeof(NumericDigit) - 1) / sizeof(NumericDigit);
+		int ndigits2 = (nbytes2 + sizeof(NumericDigit) - 1) / sizeof(NumericDigit);

 		if (nbytes1 == 1)
 		{
@@ -2750,7 +2752,7 @@ hash_numeric(PG_FUNCTION_ARGS)
 	 */
 	disk_digits = NUMERIC_DIGITS(key);
 	nbytes = NUMERIC_NBYTES(key);
-	ndigits = (nbytes + 1) / sizeof(NumericDigit);
+	ndigits = (nbytes + sizeof(NumericDigit) - 1) / sizeof(NumericDigit);
 	if (nbytes == 1)
 	{
 		fixed_buf[0] = 0;
@@ -2842,7 +2844,7 @@ hash_numeric_extended(PG_FUNCTION_ARGS)

 	disk_digits = NUMERIC_DIGITS(key);
 	nbytes = NUMERIC_NBYTES(key);
-	ndigits = (nbytes + 1) / sizeof(NumericDigit);
+	ndigits = (nbytes + sizeof(NumericDigit) - 1) / sizeof(NumericDigit);
 	if (nbytes == 1)
 	{
 		fixed_buf[0] = 0;
@@ -6846,17 +6848,26 @@ int2int4_sum(PG_FUNCTION_ARGS)
 static void
 dump_numeric(const char *str, Numeric num)
 {
-	NumericDigitData *disk_digits;
+	NumericDigitData *digit_data;
 	NumericDigit *digits;
+	int			nbytes;
 	int			ndigits;
 	int			i;

-	disk_digits = NUMERIC_DIGITS(num);
+	digit_data = NUMERIC_DIGITS(num);
+	nbytes = NUMERIC_NBYTES(num);
 	ndigits = NUMERIC_NDIGITS(num);
 	digits = digitbuf_alloc(ndigits);
-	for (i = 0; i < ndigits; i++)
+	if (nbytes == 1)
 	{
-		digits[i] = (NumericDigit) *((int16 *) &disk_digits[i*2]);
+		digits[0] = (NumericDigit) digit_data[0];
+	}
+	else
+	{
+		for (i = 0; i < ndigits; i++)
+		{
+			digits[i] = (NumericDigit) *((NumericDigit *) &digit_data[i*2]);
+		}
 	}

 	printf("%s: NUMERIC w=%d d=%d ", str,
@@ -6886,6 +6897,7 @@ dump_numeric(const char *str, Numeric num)
 	for (i = 0; i < ndigits; i++)
 		printf(" %0*d", DEC_DIGITS, digits[i]);
 	printf("\n");
+	digitbuf_free(digits);
 }


@@ -7425,7 +7437,7 @@ set_var_from_num(Numeric num, NumericVar *dest)
 {
 	int nbytes = NUMERIC_NBYTES(num);

-	alloc_var(dest, (nbytes + 1) / sizeof(NumericDigit));
+	alloc_var(dest, (nbytes + sizeof(NumericDigit) - 1) / sizeof(NumericDigit));

 	dest->weight = NUMERIC_WEIGHT(num);
 	dest->sign = NUMERIC_SIGN(num);
@@ -7470,7 +7482,7 @@ init_var_from_num(Numeric num, NumericVar *dest)
 	}
 	else
 	{
-		dest->ndigits = (nbytes + 1) / sizeof(NumericDigit);
+		dest->ndigits = (nbytes + sizeof(NumericDigit) - 1) / sizeof(NumericDigit);
 		dest->digits = (NumericDigit *) NUMERIC_DIGITS(num);
 	}

@@ -7752,7 +7764,7 @@ numericvar_serialize(StringInfo buf, const NumericVar *var)
 	pq_sendint32(buf, var->sign);
 	pq_sendint32(buf, var->dscale);
 	for (i = 0; i < var->ndigits; i++)
-		pq_sendint16(buf, var->digits[i]);
+		pq_sendint32(buf, var->digits[i]);
 }

 /*
@@ -7772,7 +7784,7 @@ numericvar_deserialize(StringInfo buf, NumericVar *var)
 	var->sign = pq_getmsgint(buf, sizeof(int32));
 	var->dscale = pq_getmsgint(buf, sizeof(int32));
 	for (i = 0; i < len; i++)
-		var->digits[i] = pq_getmsgint(buf, sizeof(int16));
+		var->digits[i] = pq_getmsgint(buf, sizeof(int32));
 }


@@ -12087,10 +12099,10 @@ accum_sum_final(NumericSumAccum *accum, NumericVar *result)
 	for (i = 0; i < accum->ndigits; i++)
 	{
 		Assert(accum->pos_digits[i] < NBASE);
-		pos_var.digits[i] = (int16) accum->pos_digits[i];
+		pos_var.digits[i] = accum->pos_digits[i];

 		Assert(accum->neg_digits[i] < NBASE);
-		neg_var.digits[i] = (int16) accum->neg_digits[i];
+		neg_var.digits[i] = accum->neg_digits[i];
 	}

 	/* And add them together */
diff --git a/src/test/regress/expected/numeric.out b/src/test/regress/expected/numeric.out
index 4731ad3fa2..da33bf1150 100644
--- a/src/test/regress/expected/numeric.out
+++ b/src/test/regress/expected/numeric.out
@@ -3558,13 +3558,13 @@ SELECT regexp_replace(
 );
      regexp_replace
 -------------------------
- 07 00 80 09 00 80 01 09+
  00 80 02 09 00 80 03 09+
- 00 80 ff 0b 00 80 00 01+
- 0b 00 80 0f 27 13 02 80+
- ab 00 cd 00 ef 00 09 00+
- 80 04 09 00 80 05 09 00+
- 80 06 09 00 80 07 00 00+
+ 00 80 ff 0f 00 80 00 01+
+ 00 00 0f 00 80 0f 27 00+
+ 00 1f 02 80 ab 00 00 00+
+ cd 00 00 00 ef 00 00 00+
+ 09 00 80 04 09 00 80 05+
+ 09 00 80 06 09 00 80 07+

 (1 row)

--
2.37.1 (Apple Git-137.1)